### PR TITLE
Required workspaces in list_objects, refactor parameters

### PR DESCRIPTION
**Pull request by MrCreosote**
This PR turned out to be really big. Sorry about that.

It does two major things, which wound up having a number of knock on effects:

1. It refactors the list object parameters class to use a builder rather than a zillion constructors.
    1. It also modernized the class a bit, using Optional and Instant instead of nulls and Dates.
2. It requires between 1 and 10000 workspaces as inputs when listing objects. Type only lists are dangerous because a large number of workspaces (potentially >> 10000) may need to be examined as part of the operation.

It didn't really make sense to not combine these two things, since refactoring the class means redoing the constructors, which is what the workspace requirements would do.

That has some knock on effects:
1. the minimum permission parameter for list objects is now useless, since a workspace list is required, so the user should just filter there.
2. likewise, the exclude global workspaces command is also useless.
3. Some other more minor changes in the part of the codebase  surrounding the parameters class.

### CODE REVIEW

File: src/us/kbase/workspace/test/workspace/WorkspaceTest.java [hunk 1]
```diff
@@ -605,68 +605,27 @@ public void adminListObjects() throws Exception {
...
 
-		lop = new ListObjectsParameters(Arrays.asList(wsi, wsi2), SAFE_TYPE1)
-				.withIncludeMetaData(true).withShowDeleted(true);
+		lop = ListObjectsParameters.getBuilder(Arrays.asList(wsi, wsi2)).withType(SAFE_TYPE1)
+				.withIncludeMetaData(true).withShowDeleted(true).withAsAdmin(true).build();
 		compareObjectInfo(ws.listObjects(lop), Arrays.asList(del, std2));
 	}
 	
-	@Test
-	public void adminListObjectsFailConstructParams() {
```
Comment (MrCreosote) [inline src/us/kbase/workspace/test/workspace/WorkspaceTest.java:629]: Made redundant by the new unit tests for the class

File: src/us/kbase/workspace/test/workspace/WorkspaceTest.java [hunk 2]
```diff
@@ -5890,64 +5851,32 @@ public void listObjectsWithDeletedObjects() throws Exception {
...
-		compareObjectInfo(ws.listObjects(lop.withShowOnlyDeleted(false)), Arrays.asList(std));
-		compareObjectInfo(ws.listObjects(lop.withShowDeleted(true)),
-				Arrays.asList(std, del));
-		compareObjectInfo(ws.listObjects(lop.withShowDeleted(false).withShowOnlyDeleted(true)),
-				Arrays.asList(del));
-		
-	}
-	
-	@Test
-	public void listObjectsWithDeletedWorkspace() throws Exception {
```
Comment (MrCreosote) [inline src/us/kbase/workspace/test/workspace/WorkspaceTest.java:5922]: This scenario can no longer occur

File: src/us/kbase/workspace/test/workspace/WorkspaceTest.java [hunk 3]
```diff
@@ -6076,224 +6009,185 @@ public void listObjectsAndHistory() throws Exception {
...
 				Arrays.asList(stdws2, hidden, deleted));
-		compareObjectInfo(ws.listObjects(new ListObjectsParameters(user2, allType2)
-				.withShowHidden(true).withShowDeleted(true)
-				.withShowAllVersions(true).withIncludeMetaData(true)),
+		compareObjectInfo(ws.listObjects(ListObjectsParameters.getBuilder(allws2)
+				.withUser(user2).withType(allType2).withShowHidden(true).withShowDeleted(true)
+				.withShowAllVersions(true).withIncludeMetaData(true).build()),
 				new ArrayList<ObjectInformation>());
 		
-		//TODO MOVE move these to unit tests for LOP
```
Comment (MrCreosote) [inline src/us/kbase/workspace/test/workspace/WorkspaceTest.java:6264]: Made redundant by the new unit tests for the class

**Comment by MrCreosote** [comment]
Travis didn't pick up the PR it seems; closing & reopening fixed it

File: lib/Bio/KBase/workspace/Client.pm [hunk 4]
```diff
@@ -10921,21 +10921,21 @@ auth has a value which is a string
 
 Parameters for the 'list_objects' function.
 
-                At least one of the following filters must be provided. It is strongly
-                recommended that the list is restricted to the workspaces of interest,
-                or the results may be very large:
+                At least one, and no more than 10000, workspaces musts be specified in one of the
```
Comment (sychan) [inline lib/Bio/KBase/workspace/Client.pm:10924]: "musts"? Is Gollum writing your docs?

File: docsource/releasenotes.rst [hunk 5]
```diff
@@ -6,6 +6,10 @@ VERSION: 0.12.0 (Released TBD)
 
 BACKWARDS INCOMPATIBILITIES:
 
+* ``list_objects`` now requires that at least one and no more than 10000 workspaces are
+  specified.
+* As such, the ``excludeGlobal`` and ``perm`` parameters no longer have any use and are
```
Comment (sychan) [inline docsource/releasenotes.rst:11]: Are they just being silently ignored? I didn't find anything in the PR that seems to generates errors or warnings for the deprecated params.

File: lib/Bio/KBase/workspace/Client.pm (hunk 4 above)
Comment (sychan) [reply lib/Bio/KBase/workspace/Client.pm:10924]: Also, what happens if someone puts in 10001 workspaces? Is the 10k limit enforced on the object type, or does the method throw an error?

Comment (MrCreosote) [reply lib/Bio/KBase/workspace/Client.pm]: > "musts"? Is Gollum writing your docs?
I actually had this fixed in the spec but forgot to recompile. Anyway, fixed

File: docsource/releasenotes.rst (hunk 5 above)
Comment (MrCreosote) [reply docsource/releasenotes.rst:11]: Yes, silently ignored. I don't want to remove them and have code start failing unnecessarily. There's no way to return warnings to the user without backwards incompatible changes to the API, unfortunately. I could log the usage, but I'm not sure what we'd do with those logs

**Pull request merged**